client/taskenv/env.go

```diff
@@ -476,15 +476,25 @@ func (b *Builder) Build() *TaskEnv {
 	// Clean keys (see #2405)
 	cleanedEnv := make(map[string]string, len(envMap))
 	for k, v := range envMap {
-		cleanedK := helper.CleanEnvVar(k, '_')
+		var cleanedK string
+		if strings.HasPrefix(k, "NOMAD_") {
+			cleanedK = helper.CleanEnvVar(k, '_')
+		} else {
+			cleanedK = k
+		}
 		cleanedEnv[cleanedK] = v
 	}
 
 	var cleanedDeviceEnvs map[string]string
 	if deviceEnvs != nil {
 		cleanedDeviceEnvs = make(map[string]string, len(deviceEnvs))
 		for k, v := range deviceEnvs {
-			cleanedK := helper.CleanEnvVar(k, '_')
+			var cleanedK string
+			if strings.HasPrefix(k, "NOMAD_") {
```

Comment:
I think the environment variables set by device plugins should be treated like user variables and be passed varbatim; as they aren't generated by Nomad.